man/last.Rd

```diff
@@ -3,36 +3,47 @@
 \alias{last}
 \title{ First/last item of an object }
 \description{
-Returns the first/last item of a vector or list, or the first/last row of a data.frame
-or data.table. The main difference to head/tail is that the default for \code{n} is 1
-rather than 6.
+Returns the first/last item of a vector, list, \code{data.frame} or \code{data.table}. The main difference
+to head/tail is that the default for \code{n} is 1 rather than 6.
 }
 \usage{
-first(x, n=1L, \dots)
-last(x, n=1L, \dots)
+first(x, n=1L, na.rm=FALSE, \dots)
+last(x, n=1L, na.rm=FALSE, \dots)
 }
 \arguments{
-\item{x}{ A vector, list, data.frame or data.table. Otherwise the S3 method
-of \code{xts::first} is deployed. }
+\item{x}{ A vector, list, \code{data.frame} or \code{data.table}. \code{xts} objects dispatch to the \code{xts} method. }
 \item{n}{ A numeric vector length 1. How many items to select. }
+\item{na.rm}{ \code{FALSE}, \code{TRUE} or \code{"row"}. Should missing values be skipped when selecting values? Selects always at least 1 element. }
 \item{\dots}{ Not applicable for \code{data.table} first/last. Any arguments here
 are passed through to \code{xts}'s first/last. }
 }
 \value{
-If no other arguments are supplied it depends on the type of \code{x}. The first/last item
-of a vector or list. The first/last row of a \code{data.frame} or \code{data.table}.
-For other types, or if any argument is supplied in addition to \code{x} (such as \code{n}, or
-\code{keep} in \code{xts}) regardless of \code{x}'s type, then \code{xts::first}/
-\code{xts::last} is called if \code{xts} has been loaded, otherwise \code{utils::head}/\code{utils::tail}.
+The first/last \code{n} items of a vector or list.
```

Comment:
Should mention the padding with NA.

Reply:
I'm a bit uncomfortable about the padding. The reason for it was because we don't have a way for GForce to tell the `[.data.table` caller how many items are in each group where that differs from `min(n,grpsize)` due to how many non-NA there are. If it wasn't for that then we'd return the first/last `n` non-NA as naturally expected I assume we all agree? If so then it's just our internals that are driving the user interface on this padding currently.
The trouble is, if we go ahead with the padding, then we'll have to keep that in future for backwards compatibility, and we'll have to add a controlling argument to provide the non-padding if and when implemented. I can't imagine anyone really wants the padding.
Shall I see how hard it might be to add that vector of the result size for each group to be returned from GForce? Might be useful in other functions too. Frustrating to still not finish this PR but now would be the time to get this right.

Reply:
True. Regarding speed we would have to iterate over the whole set twice? However, we would also be able to remove groups which do not contain non NA values which would quite useful, I guess.

What if we allocate enough space for `min(n, grpsize)` and then make a call to `realloc`?

Reply:
Moving more group information into `gsumm.c:gforce` would not only benefit `gfirstlast` but is probably helpful/necessary for other future gforce functions such as `gunique`. (At least I had to do it on my last try on `gunique` which tried to resolve `gunique` with a single sort and kept track of whether elements are increasing per group).

Reply:
Now done and called it `gforce_dynamic`. Currently `gfirstlast` and `gshift` attach this attribute (all other `gfuns` return one value per group). `gsumm.c:gforce` then works out the union shape of the potentially many gforce dynamic columns, aligns them, and reps the grpcols appropriately. Feels right now that the `gfuns` return their shape rather than having to code knowledge of the `gfuns` at R level.